### PR TITLE
Fix race condition on proxy server

### LINKED ISSUE #373: Proxy server failure on multiple calls: starknet-venv-proxy exited with code=1 while running
As reported [here](https://discord.com/channels/793094838509764618/912735106899275856/1107984098825343087) by a user.

### PR DESCRIPTION
**Pull request by Nathan-SL**
## Usage related changes

-   Improves the behavior of the proxy server when making multiple calls to fetch pre-deployed accounts.
-   Closes #373

## Development related changes

-  Adds a function to ensure proper assignment of a free port.

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors + tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [x] Rebased to the last commit of the target branch (or merged it into my branch)
-   [ ] Documented the changes
-   [x] Updated the `test` directory (with a test case consisting of `network.json`, `hardhat.config.ts`, `check.ts`)
-   [x] Linked issues which this PR resolves
-   [x] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/0xSpaceShard/starknet-hardhat-example):https://github.com/0xSpaceShard/starknet-hardhat-example/pull/112
    -   [x] Modified `test.sh` to use my example repo branch
    -   [ ] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [x] All tests are passing (for external contributors who don't have access to the CI/CD pipeline)

### CODE REVIEW

File: src/external-server/external-server.ts [hunk 1]
```diff
@@ -177,8 +179,10 @@ export abstract class ExternalServer {
     }
 
     public async post<T>(data: StringMap): Promise<T> {
+        await this.ensurePort();
         await this.ensureStarted();
 
+        axiosRetry(axios, { retries: 5, retryDelay: axiosRetry.exponentialDelay });
```
Comment (FabijanC) [inline src/external-server/external-server.ts:185]: Add a comment here about it not working in CircleCI with `retries: 3`. Can you find out how much time these 5 retries actually amount to?

File: test/integrated-devnet-tests/predeployed-accounts-test/check.ts [hunk 2]
```diff
@@ -0,0 +1,8 @@
+import { hardhatStarknetTest } from "../../utils/cli-functions";
+import { checkDevnetIsNotRunning } from "../../utils/utils";
+
```
Comment (FabijanC) [inline test/integrated-devnet-tests/predeployed-accounts-test/check.ts:3]: Add a comment here describing what we are testing and why this is in integrated-devnet tests.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: test/integrated-devnet-tests/predeployed-accounts-test/check.ts (hunk 2 above)
Comment (Nathan-SL) [reply test/integrated-devnet-tests/predeployed-accounts-test/check.ts:3]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/374/commits/e7b25f93f7cf0ee5428958cf1b3e7469586a624d

File: src/external-server/external-server.ts (hunk 1 above)
Comment (Nathan-SL) [reply src/external-server/external-server.ts:186]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/374/commits/e7b25f93f7cf0ee5428958cf1b3e7469586a624d

**Pull request merged**